src/controller-v5.sol

```diff
@@ -0,0 +1,375 @@
+    }
+
+    function inCaseStrategyTokenGetStuck(address _strategy, address _token)
+        public
+    {
+        require(
+            msg.sender == strategist || msg.sender == governance,
+            "!governance"
+        );
+        IStrategy(_strategy).withdraw(_token);
+    }
+
+    function getExpectedReturn(
+        address _strategy,
+        address _token,
+        uint256 parts
+    ) public view returns (uint256 expected) {
+        uint256 _balance = IERC20(_token).balanceOf(_strategy);
+        address _want = IStrategy(_strategy).want();
+        (expected, ) = OneSplitAudit(onesplit).getExpectedReturn(
+            _token,
+            _want,
+            _balance,
+            parts,
+            0
+        );
+    }
+
+    // Only allows to withdraw non-core strategy tokens ~ this is over and above normal yield
+    function yearn(
+        address _strategy,
+        address _token,
+        uint256 parts
+    ) public {
+        require(
+            msg.sender == strategist || msg.sender == governance,
+            "!governance"
+        );
+        // This contract should never have value in it, but just incase since this is a public call
+        uint256 _before = IERC20(_token).balanceOf(address(this));
+        IStrategy(_strategy).withdraw(_token);
+        uint256 _after = IERC20(_token).balanceOf(address(this));
+        if (_after > _before) {
+            uint256 _amount = _after.sub(_before);
+            address _want = IStrategy(_strategy).want();
+            uint256[] memory _distribution;
+            uint256 _expected;
+            _before = IERC20(_want).balanceOf(address(this));
+            IERC20(_token).safeApprove(onesplit, 0);
+            IERC20(_token).safeApprove(onesplit, _amount);
+            (_expected, _distribution) = OneSplitAudit(onesplit)
+                .getExpectedReturn(_token, _want, _amount, parts, 0);
+            OneSplitAudit(onesplit).swap(
+                _token,
+                _want,
+                _amount,
+                _expected,
+                _distribution,
+                0
+            );
+            _after = IERC20(_want).balanceOf(address(this));
+            if (_after > _before) {
+                _amount = _after.sub(_before);
+                uint256 _treasury = _amount.mul(split).div(max);
+                earn(_want, _amount.sub(_treasury));
+                IERC20(_want).safeTransfer(treasury, _treasury);
+            }
+        }
+    }
+
+    function withdraw(address _token, uint256 _amount) public {
+        require(msg.sender == jars[_token], "!jar");
+        IStrategy(strategies[_token]).withdraw(_amount);
+    }
+
+    function withdrawReward(address _token, uint256 _reward) public {
+        require(msg.sender == jars[_token], "!jar");
+        IStrategy(strategies[_token]).withdrawReward(_reward);
+    }
+
+    // Function to swap between jars
+    function swapExactJarForJar(
+        address _fromJar, // From which Jar
```

Comment:
I remember one of the root cause of the incident with the controller was not checking that these provided Jars are known in the Pickle Universe (validated) is that check done somewhere else? 

Or how do we handle someone sending malicious jar? i don't remember the fix for that incident.